neffint/adaptive_fourier_integral.py

```diff
@@ -0,0 +1,465 @@
+    Results will be cached to avoid repeat calls with the same input.
+    :type func: Callable[[float], ArrayLike]
+    :param interpolation_error_metric: A callable that takes in an array of func outputs (shape (N, X1, X2, ...)) and interpolated approximations
+    of the same values (also shape (N, X1, X2, ...))
+    The primary axis (of length N) corresponds to the frequency. The callable should return a difference metric between the two inputs,
+    collapsed into a 1D array along the frequency axis,
+    i.e. the output should have the shape (N,)
+    :type interpolation_error_metric: Callable[[np.ndarray, np.ndarray], np.ndarray]
+    :param absolute_integral_tolerance: The tolerance for the (approximate) integrated interpolation error. When the error drops below this value, the algorithm terminates.
+    :type absolute_integral_tolerance: float
+    :param logstep_towards_inf: When one of the ends of the frequency array is +-inf, the bisection of the corresponding interval is calculated by multiplying the (finite)
+    neighboring frequency with this number. Defaults to 2
+    :type logstep_towards_inf: float, optional
+    :param bisection_mode_condition: A function that takes in a frequency array [Hz]* and returns an array of bools of the same size,
+    where `True` denotes that the interval following that frequency should be bisected linearly,
+    and `False` geometrically. Can be set to `None`, in which case only geometric bisection is used. Defaults to `None`
+    :type bisection_mode_condition: Optional[Callable[[np.ndarray], np.ndarray]], optional
+    :param max_iterations: The maximum amount of iterations to perform before the algorithm is terminated. If set to `None`, the algorithm will only terminate
+    after the tolerance has been met. Defaults to `None`
+    :type max_iterations: Optional[int], optional
+    :return: The refined frequency range [Hz]*, and an array containing the corresponding outputs of `func`
+    :rtype: Tuple[np.ndarray, np.ndarray]
+    
+    * Though Hz is used as units here, any frequency unit will work, just be mindful that the unit must be coherent with the time unit if used in a Fourier integral
+    """
+    
+    # NOTE: Could consider changing all insert and append to work in-place on a larger array. I.e. a dynamic array approach
+
+    # Starting frequencies
+    frequencies = np.asarray(initial_frequencies)
+    assert len(frequencies) >= 2, "Need 2 or more frequencies in initial frequency range"
+    
+    # To avoid algorithmic problems, insert `-1` if the first frequency is -inf and the second is non-negative
+    if frequencies[0] == -np.inf and frequencies[1] >= 0:
+        frequencies = np.insert(frequencies, 1, -1)
+    
+    # Similarly, if the last frequency is +inf and the second last is non-positive, insert the frequency `1`
+    if frequencies[-1] == np.inf and frequencies[-2] <= 0:
+        frequencies = np.insert(frequencies, -1, 1)
```

Comment:
It must be a finite *non-zero* number of the same sign as the infinity, since the new frequencies are made by multiplying the finite number with a constant, which of course would not work for 0. I agree that 1 is quite arbitrary, the thought was that the algorithm would hopefully fairly quickly find the important frequencies to add, anyway.